witness/src/Dscp/Witness/Retranslate/Transaction.hs

```diff
@@ -0,0 +1,116 @@
+--      These two cannot be put into one 'STM' action, sadly, because `cliRecvUpdate`
+--      is not 'STM' and reading bytestring from `btq` directly is not what we want
+--      on this abstraction level.
+--
+--   2) Each of these workers receives txs to retranslate from its respective source,
+--      checks that they are applicable to retranslation and pushes them into the pipe.
+--
+--   3) Republisher drains the pipe and publishes everything that comes from it.
+--
+--   This is the connection scheme of the retranslator:
+--
+--   other 'PubTx's --> trSubConsumer --+
+--                                      |
+--                                      V
+--   input --> trMessageConsumer --> pipe --> trRepublisher
+--
+makeRetranslator :: forall ctx m. WitnessWorkMode ctx m => STM (TxRetranslator m)
+makeRetranslator = do
+    failedTxs <- STM.newTVar HashMap.empty
+    input     <- STM.newTQueue
+    pipe      <- STM.newTQueue
+
+    return (retranslator failedTxs input pipe)
+  where
+    retranslator failedTxs input pipe =  TxRetranslator
+        {
+          trInput = input
+
+        , trMessageConsumer = Worker
+            "txRetranslationInitialiser"
+            [msgType @SendTx]
+            [subType @PubTx] $ \_btq -> do
+                dieGracefully $ do
+                    forever $ do
+                        SendTx tx <- sync $ STM.readTQueue input
+                        checkThenRepublish tx
+
+        , trSubConsumer = Worker
+            "txRetranslationRepeater"
+            [msgType @SendTx]
+            [subType @PubTx] $ \btq -> do
+                dieGracefully $ do
+                    forever $ do
+                        (_, PubTx tx) <- cliRecvUpdate btq (-1)
+                        checkThenRepublish tx
+
+        , trRepublisher = Listener
+            "txRetranslationPublisher"
+            [] $ \btq -> do
+                dieGracefully $ do
+                    forever $ sync $ do
+                        tx <- STM.readTQueue pipe
+                        servPub btq (PubTx tx)
+        }
+      where
+        -- Check that the transaction is neither failed nor in a mempool and republish it.
+        checkThenRepublish tx = do
+            hashmap <- sync $ readTVar failedTxs
+
+            unless (hash tx `elem` hashmap) $ do
+                pool    <- view (lensOf @MempoolVar @ctx)
+                isThere <- isInMempool pool tx
+
+                unless isThere $ do
+                    addTxToMempool pool tx
+                    sync $ STM.writeTQueue pipe tx
+
+        -- Catch error and log it.
+        -- TODO: move to some *.Utils.
+        dieGracefully action =
+            action `catchAny` \e -> do
+                logError $ fromString $ "Exception in transactionRetranslatorInput: " <> show e
+
+-- | Run STM action inside (MonadIO m).
+--   TODO: move to some *.Utils.
+sync :: MonadIO m => STM.STM a -> m a
+sync = liftIO . STM.atomically
```

Comment:
There is `atomically` in `universum` which does the same

Reply:
Nice. However, I don't want to drop `STM.` prefix from them here. It specifies that they are from STM module. In `Universum` case, they are just imported silently.

Reply:
Please do it, everybody knows what `atomically` does. In general, if something is present in universum, it should be prioritised over direct imports.